veritysetup: add support for veritytab

gportay opened the pull request:
Hello,

This PR brings support for veritytab in the given format:

`<volume-name> <data-device> <hash-device> <root-hash> <coma-separated-options>`

The veritytab file contains at most five fields, the first four are mandatory, the last one is optional:
- The first field contains the name of the resulting verity volume; its block device is set up `/dev/mapper/</filename>`.
- The second field contains a path to the underlying block data device, or a specification of a block device via _UUID=_ followed by the UUID.
- The third field contains a path to the underlying block hash device, or a specification of a block device via _UUID=_ followed by the UUID.
- The fourth field is the roothash in hexadecimal.
- The fifth field, if present, is a comma-delimited list of options.

For now, the following options are recognized only: `ignore-corruption`, `restart-on-corruption`, `panic-on-corruption`, `ignore-zero-blocks`, `check-at-most-once`. The others options will be implemented later.

Initially, these changes were requested to solve issue below.

I use `bmaptool` to copy a disk image to a physical device. That utility does not write the blocks that are unused in order to speed up the copy.

Therefore, the unused blocks are not filled with zero and the verity hashes for these blocks won't match.

The `cryptsetup` library has the flag `CRYPT_ACTIVATE_IGNORE_ZERO_BLOCKS` that instructs the kernel to not verify blocks that are expected to contain zeroes and always directly return zeroes instead. This flag is enabled by the option `--ignore-zero-blocks` in `veritysetup(8)`.

More generally, the `cryptsetup` library has 5 different flags that alter the behaviour and this PR allows controlling it by setting ~~the appropriate environment variable to true or false~~ the kernel command-line argument `systemd.verity.options=OPTION[|OPTION...]`.

~~It introduces the 5 environment variables to `systemd-veritysetup(8)`, each variable controls a specific `dm-verity` flag; the variables are prefixed by `SYSTEMD_` and are named as the library's flag:~~

- ~~`$SYSTEMD_CRYPT_ACTIVATE_IGNORE_CORRUPTION=1`~~ `ignore-corruption` — Without this option kernel fails the IO operation with I/O error. If set, the corruption is only logged.
- ~~`$SYSTEMD_CRYPT_ACTIVATE_RESTART_ON_CORRUPTION=1`~~ `restart-on-corruption` — Without this option kernel fails the IO operation with I/O error. If set, the kernel is restarted immediately.
- ~~`$SYSTEMD_CRYPT_ACTIVATE_IGNORE_ZERO_BLOCKS=1`~~ `ignore-zero-blocks` — Instruct kernel to not verify blocks that are expected to contain zeroes and always directly return zeroes instead.
- ~~`$SYSTEMD_CRYPT_ACTIVATE_CHECK_AT_MOST_ONCE=1`~~ `check-at-most-once` — Instruct kernel to verify blocks only the first time they are read from the data device, rather than every time.
- ~~`$SYSTEMD_CRYPT_ACTIVATE_PANIC_ON_CORRUPTION=1`~~ `panic-on-corruption` — Without this option kernel fails the IO operation with I/O error. If set, the kernel panic immediatly.

See `veritysetup(8)` for more details.

Regards,
Gaël

bluca commented on src/veritysetup/veritysetup.c:
```diff
@@ -89,6 +91,31 @@ static int run(int argc, char *argv[]) {
+                        flags |= CRYPT_ACTIVATE_RESTART_ON_CORRUPTION;
+#endif
+
+#ifdef CRYPT_ACTIVATE_IGNORE_ZERO_BLOCKS
```
The minimum libcryptsetup supported version is 2.0.1, and the following options were supported before that, so these 3 ifdefs can be removed:

CRYPT_ACTIVATE_IGNORE_CORRUPTION
CRYPT_ACTIVATE_RESTART_ON_CORRUPTION
CRYPT_ACTIVATE_IGNORE_ZERO_BLOCKS

https://gitlab.com/cryptsetup/cryptsetup/-/commit/4dd703ea6c49881a

gportay replied:
Is this a mandatory change? I prefer to keep these for consistency; anyway, I can remove these and keep both `CRYPT_ACTIVATE_CHECK_AT_MOST_ONCE` and `CRYPT_ACTIVATE_PANIC_ON_CORRUPTION`.

bluca replied:
I'm not a maintainer, so I can't mandate anything :-) But in the past, maintainers have consistently expressed a preference for keeping \#ifdefs down to a bare minimum, and for unconditionally using interfaces included in minimum supported versions of external dependencies.

gportay replied:
Thanks, I will remove them.

poettering commented:
Hmm, I'd try to keep this as close as possible to the /etc/crypttab and cryptsetup behaviour. i.e. I think systemd-veritysetup should take an "options" parameter in addition to the existing parameters, that take this kind of stuff, and eventually that should be sourced from a new column in a file /etc/verityttab that systemd-veritysetup-generator should read, as well as from the kernel cmdline, specifically an option `verity.options=` or so.

gportay commented:
> Hmm, I'd try to keep this as close as possible to the /etc/crypttab and cryptsetup behaviour. i.e. I think systemd-veritysetup should take an "options" parameter in addition to the existing parameters, that take this kind of stuff, and eventually that should be sourced from a new column in a file /etc/verityttab that systemd-veritysetup-generator should read, as well as from the kernel cmdline, specifically an option `verity.options=` or so.

If I get it right, (correct me if I am wrong), there are three things to do:

1. First, implement new "options" in the interface of `systemd-verity-setup(8)`, which concerns the `attach` action only.
Something like `systemd-verity-setup attach VOLUME DATADEVICE HASHDEVICE ROOTHASH [ROOTHASHSIG | - ] ignore-zero-blocks,restart-on-corruption,...` (the changes will be simple).
Or something more like `systemd-verity-setup attach --options ignore-zero-blocks,restart-on-corruption,... VOLUME DATADEVICE HASHDEVICE ROOTHASH [ROOTHASHSIG]` (the changes need getopt_long).
Or something like `systemd-verity-setup attach --options ignore-zero-blocks,restart-on-corruption,... VOLUME DATADEVICE HASHDEVICE ROOTHASH [ROOTHASHSIG]` (the changes need getopt_long and more code in the generator).

2. Then, implement this in the `systemd-veritysetup-generator(8)` to add that new "options" to the line `ExecStart=`.
`systemd-veritysetup-generator(8)` will read that "options" from:
- either the new kernel command-line `verity.options=` or `systemd.verity.options=` (as `roothash=`, `systemd.verity_root_data=`, `systemd.verity_root_hash=`).
- or from a `/etc/` tabfile (see 3.).

3. But `systemd-veritysetup-generator(8)` does not support yet reading information from a `/etc` tabfile.
The verity devices get attached thanks to three different parameters (data-device, hash-device and roothash) while standard cryptsetup devices get created thanks to one single parameter (encrypted-device). The file `/etc/crypttab` holds the two mandatory `volume-name`, the `encrypted-device` parameters plus the two optional `key-file` and `options` parameters. This is not enough for describing a verity device as it needs two more columns for the parameters `hash-device` and `roothash`.
Therefore, better to have a `/etc/veritytab` than recycling `/etc/crypttab` for verity (which would been weird to mix two different things anyway). [edit: I see you mentionned about a `/etc/veritytab` and not a `/etc/crypttab` after I re-read your comment; I see a reference in the code as well.]

The point 3. could be adressed in a separate PR.

bluca commented on src/veritysetup/veritysetup.c:
```diff
@@ -81,6 +82,35 @@ static int run(int argc, char *argv[]) {
                         return 0;
                 }
 
+                if (argc > 7 && strcmp(argv[7], "-")) {
```
nit: !streq is generally preferred, as it's more readable

gportay replied:
Good to know. Thanks.

bluca commented on src/veritysetup/veritysetup.c:
```diff
@@ -31,7 +31,7 @@ static int help(void) {
                 return log_oom();
 
-        printf("%s attach VOLUME DATADEVICE HASHDEVICE ROOTHASH [ROOTHASHSIG]\n"
+        printf("%s attach VOLUME DATADEVICE HASHDEVICE ROOTHASH [ROOTHASHSIG] [OPTIONS]\n"
```
I think the fact you can pass options and omit roothashsig by using "-" needs to be documented here, as it's not obvious

gportay replied:
Indeed. I would like to know if this would be possible to bypass `ROOTHASHSIG` using `-`. `-` may mean read from stdin; that is not the case here. Personnaly I find this uggly to have two optional parameters, it is the easiest way to implement generator side :) I am not against implementing options in the form `--long-option-name`, it will requires more code in the generator and I am still discovering how things work.

bluca replied:
Given there are multiple verity options supported, continuing this way essentially means we can't extend it anymore in the future, should the need manifest. So I'd be very much in favour of adding --log-option-name support. Would you be up for it?

gportay replied:
Sure I will do it. I have planned to implement most of the options offered by setupverity.

bluca commented:
CI failures appear unrelated - usual timeout in bionic-i386, NTP failure in semaphoreci, seccomp failure in travis

poettering commented on src/veritysetup/veritysetup.c:
```diff
@@ -81,6 +82,35 @@ static int run(int argc, char *argv[]) {
+                                if (r == 0)
+                                        break;
+
+                                if (strcmp(word, "ignore-corruption") == 0)
```
please use `streq(…)` instead of `strcmp(…) == 0` in our codebase as a matter of coding style.

poettering commented on src/veritysetup/veritysetup.c:
```diff
@@ -81,6 +82,35 @@ static int run(int argc, char *argv[]) {
+                        for (;;) {
+                                _cleanup_free_ char *word = NULL;
+
+                                r = extract_first_word((const char **)&argv[7], &word, ",", EXTRACT_DONT_COALESCE_SEPARATORS | EXTRACT_UNESCAPE_SEPARATORS);
```
uh, you are modifiying the argv[] array here, that's not OK, please do this in a temporary pointer you initialize from argv[7], but do not modify argv[7] itself.

poettering commented on src/veritysetup/veritysetup-generator.c:
```diff
@@ -158,6 +164,16 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
+                r = free_and_strdup(&arg_options, value);
+                if (r < 0)
+                        return log_oom();
+
```
doesn#t matter, but this empty line seems too much ;-)

gportay replied:
Sure, tbh, I have added that line because that same empty line exists in the other if statement above ;)

I can drop them as well in another commit.

poettering reviewed (changes requested):
looks excellent. Some minor comments

gportay commented:
I will implement all the options using --long-option (whenever I will find the time to...)

gportay commented:
And I will implement `veritytab` support.

gportay commented:
I have rebased on top of master and added an early implementation for veritytab. Most of the work is greatly inspired on `cryptsetup-generator`.

Reviews are welcome.

bluca commented:
Looks fine on first look, but veritytab is missing support for root hash signatures - could you please add that too? Otherwise there's a mismatch

gportay commented:
> Looks fine on first look, but veritytab is missing support for root hash signatures - could you please add that too? Otherwise there's a mismatch

Sure. It is in my todo list.

gportay commented:
TBH, I am thinking about replacing the positional argument `ROOTHASHSIG` with option `root-hash-signature=(file|base64:sig)` which would be the equivalent of `--root-hash-signature` in the world of cryptsetup/veritysetup.

What do you think about that change? It should not break the existing behaviour I guess.

``` diff
--- a/src/veritysetup/veritysetup.c
+++ b/src/veritysetup/veritysetup.c
@@ -19,10 +19,12 @@ static char *arg_root_hash = NULL;
 static char *arg_data_what = NULL;
 static char *arg_hash_what = NULL;
 static int arg_activate_flags = CRYPT_ACTIVATE_READONLY;
+static char *arg_root_hash_signature = NULL;
 
 STATIC_DESTRUCTOR_REGISTER(arg_root_hash, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_data_what, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_hash_what, freep);
+STATIC_DESTRUCTOR_REGISTER(arg_root_hash_signature, freep);
 
 static int help(void) {
         _cleanup_free_ char *link = NULL;
@@ -32,7 +34,7 @@ static int help(void) {
         if (r < 0)
                 return log_oom();
 
-        printf("%s attach VOLUME DATADEVICE HASHDEVICE ROOTHASH [ROOTHASHSIG] [OPTIONS]\n"
+        printf("%s attach VOLUME DATADEVICE HASHDEVICE ROOTHASH [OPTIONS]\n"
                "%s detach VOLUME\n\n"
                "Attaches or detaches an integrity protected block device.\n"
                "\nSee the %s for details.\n"
@@ -54,6 +56,8 @@ static int parse_options(const char *options) {
 
         for (;;) {
                 _cleanup_free_ char *word = NULL;
+                struct stat st;
+                char *val;
 
                 r = extract_first_word(&copy, &word, ",", EXTRACT_DONT_COALESCE_SEPARATORS | EXTRACT_UNESCAPE_SEPARATORS);
                 if (r < 0)
@@ -78,7 +82,12 @@ static int parse_options(const char *options) {
                 else if (streq(word, "panic-on-corruption"))
                         arg_activate_flags |= CRYPT_ACTIVATE_PANIC_ON_CORRUPTION;
 #endif
-                else
+                else if ((val = startswith(word, "root-hash-signature="))) {
+                        arg_root_hash_signature = val;
+                } else if (stat(word, &st) == 0 || (val = startswith(word, "base64:"))) {
+                        log_warning("Usage of ROOTHASHSIG positional argument is deprecated. Please use the option root-hash-signature=%s.", word);
+                        arg_root_hash_signature = word;
+                } else
                         return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Unknown option %s.", word);
         }
 
@@ -123,8 +132,8 @@ static int run(int argc, char *argv[]) {
                         return 0;
                 }
 
-                if (argc > 7) {
-                        r = parse_options(argv[7]);
+                if (argc > 6) {
+                        r = parse_options(argv[6]);
                         if (r < 0)
                                 return log_error_errno(r, "Failed to parse options: %m");
                 }
@@ -137,19 +146,19 @@ static int run(int argc, char *argv[]) {
                 if (r < 0)
                         return log_error_errno(r, "Failed to configure data device: %m");
 
-                if (argc > 6 && *argv[6]) {
+                if (*arg_root_hash_signature) {
 #if HAVE_CRYPT_ACTIVATE_BY_SIGNED_KEY
                         _cleanup_free_ char *hash_sig = NULL;
                         size_t hash_sig_size;
                         char *value;
 
-                        if ((value = startswith(argv[6], "base64:"))) {
+                        if ((value = startswith(arg_root_hash_signature, "base64:"))) {
                                 r = unbase64mem(value, strlen(value), (void *)&hash_sig, &hash_sig_size);
                                 if (r < 0)
-                                        return log_error_errno(r, "Failed to parse root hash signature '%s': %m", argv[6]);
+                                        return log_error_errno(r, "Failed to parse root hash signature '%s': %m", arg_root_hash_signature);
                         } else {
                                 r = read_full_file_full(
-                                                AT_FDCWD, argv[6], UINT64_MAX, SIZE_MAX,
+                                                AT_FDCWD, arg_root_hash_signature, UINT64_MAX, SIZE_MAX,
                                                 READ_FULL_FILE_CONNECT_SOCKET,
                                                 NULL,
                                                 &hash_sig, &hash_sig_size);

```
Including that change to the generator.

``` diff
--- a/src/veritysetup/veritysetup-generator.c
+++ b/src/veritysetup/veritysetup-generator.c
@@ -122,7 +122,7 @@ static int create_device(void) {
                 "\n[Service]\n"
                 "Type=oneshot\n"
                 "RemainAfterExit=yes\n"
-                "ExecStart=" ROOTLIBEXECDIR "/systemd-veritysetup attach root '%s' '%s' '%s' '' '%s'\n"
+                "ExecStart=" ROOTLIBEXECDIR "/systemd-veritysetup attach root '%s' '%s' '%s' '%s'\n"
                 "ExecStop=" ROOTLIBEXECDIR "/systemd-veritysetup detach root\n",
                 d, e,
                 d, e,
```

bluca commented:
Yes, that's a good idea, looks much better

gportay commented:
> Yes, that's a good idea, looks much better

Great! I have pushed that change but I do not know how to test it to check if I have not introduced a regression.

bluca commented on src/veritysetup/veritysetup.c:
```diff
@@ -78,7 +82,18 @@ static int parse_options(const char *options) {
+                        r = free_and_strdup(&arg_root_hash_signature, val);
+                        if (r < 0)
+                                return log_oom();
+                } else if (stat(word, &st) == 0 || (val = startswith(word, "base64:"))) {
```
nit: let's invert this check, so that we don't try and stat weird non-existing paths if base64:foo is passed

gportay replied:
Good nit indeed.

bluca commented on man/veritytab.xml:
```diff
@@ -0,0 +1,167 @@
+        corruption is only logged. With <literal>--restart-on-corruption</literal> or
+        <literal>--panic-on-corruption</literal> the kernel is restarted (panicked) immediately.
+
+        (You have to provide way how to avoid restart loops.)</para></listitem>
```
What does this sentence mean? How can a user of veritytab do that? (also /way/a way/)

gportay replied:
TBH, I took the description from the [veritysetup(8)](https://man7.org/linux/man-pages/man8/veritysetup.8.html). I guess it means you need to have a bootloader that can count boot failures and start a fallback procedure to prevent the system from rebooting in a loop.

Better to discard that sentence?

bluca replied:
That makes sense, let's leave it in then

bluca commented on man/veritytab.xml:
```diff
@@ -0,0 +1,167 @@
+
+        <listitem><para>This device will not be added to <filename>cryptsetup-pre.target</filename>.
+        This means that it will not be automatically unlocked on boot, unless something else pulls
+        it in. In particular, if the device is used for a mount point, it'll be unlocked
```
I'm not sure "unlocked" is the best word here? It's not a crypto device after all. "enabled" maybe?

gportay replied:
Oops, I forgot to reword that part. I took it from [crypttab(5)](https://www.freedesktop.org/software/systemd/man/crypttab.html).

gportay replied:
"enabled" seems fine.

bluca commented on src/veritysetup/veritysetup-generator.c:
```diff
@@ -109,7 +122,7 @@ static int create_device(void) {
                 "Type=oneshot\n"
                 "RemainAfterExit=yes\n"
-                "ExecStart=" ROOTLIBEXECDIR "/systemd-veritysetup attach root '%s' '%s' '%s' - '%s'\n"
+                "ExecStart=" ROOTLIBEXECDIR "/systemd-veritysetup attach root '%s' '%s' '%s' '' '%s'\n"
```
what's this change for?

gportay replied:
Looks like it is a rebase error.

gportay replied:
It looks like a rebase error, my bad.

bluca commented on src/veritysetup/veritysetup-generator.c:
```diff
@@ -220,21 +240,214 @@ static int determine_devices(void) {
+                const char *options,
+                const char *source) {
+
+        _cleanup_free_ char *n = NULL, *du = NULL, *hu = NULL, *e = NULL, *filtered = NULL, *du_escaped = NULL, *hu_escaped = NULL, *name_escaped = NULL;
```
nit: please break this line

gportay replied:
How long is a max line?

bluca replied:
When it feels too long in a diff :-)

```
- Don't break code lines too eagerly. We do **not** force line breaks at 80ch,
  all of today's screens should be much larger than that. But then again, don't
  overdo it, ~109ch should be enough really.
```

https://github.com/systemd/systemd/blob/master/docs/CODING_STYLE.md

gportay replied:
thanks

bluca commented on src/veritysetup/veritysetup.c:
```diff
@@ -60,19 +61,22 @@ static int parse_flags(const char *flags) {
                 if (r == 0)
                         break;
 
+                if (STR_IN_SET(word, "noauto", "auto", "nofail", "fail"))
```
Please split out changes to this file, apart from this one if block, from commit "veritysetup-generator: add support for veritytab" and squash it back into the first commit, "veritysetup: add support for dm-verity flags"

We want "perfect" commits in the git history

bluca commented on src/veritysetup/veritysetup.c:
```diff
@@ -81,6 +120,12 @@ static int run(int argc, char *argv[]) {
 
+                if (argc > 7) {
+                        flags = parse_flags(argv[7]);
+                        if (r < 0)
```
this doesn't look right - it's fixed in a later commit, and those changes should instead be squashed back into the first commit

bluca commented:
Nice! Very close, just a few more fixes.

Also, but we don't need to sort it now, I wonder if we should have new veritysetup targets instead of piggy-backing onto the cryptsetup ones. But it doesn't have to be sorted here, it can be done separately.

gportay commented:
> Also, but we don't need to sort it now, I wonder if we should have new veritysetup targets instead of piggy-backing onto the cryptsetup ones. But it doesn't have to be sorted here, it can be done separately.

I have a pending commit that brings the `veritysetup.target` but it misses the glue that pulls that target at bootup... I wonder which unit or what is the thing that should pull it up.

bluca commented:
> > Also, but we don't need to sort it now, I wonder if we should have new veritysetup targets instead of piggy-backing onto the cryptsetup ones. But it doesn't have to be sorted here, it can be done separately.
> 
> I have a pending commit that brings the `veritysetup.target` but it misses the glue that pulls that target at bootup... I wonder which unit or what is the thing that should pull it up.

sysinit.target, just like cryptsetup is

gportay commented:
> sysinit.target, just like cryptsetup is

Oh, I found what I have missed... that [line](https://github.com/systemd/systemd/blob/master/units/meson.build#L12).

gportay commented:
These new changes introduce the target `veritysetup.target` and a couple of fixes. I have also removed references to timeouts which are not necessary.

However, there is still a reference to `cryptsetup.target` if rootfs uses verity. I have not tested it yet and I will do it ASAP.

gportay commented:
I will perform some corner case tests as well.

poettering commented on src/veritysetup/veritysetup.c:
```diff
@@ -43,6 +46,59 @@ static int help(void) {
+                struct stat st;
+                char *val;
+
+                r = extract_first_word(&copy, &word, ",", EXTRACT_DONT_COALESCE_SEPARATORS | EXTRACT_UNESCAPE_SEPARATORS);
```
This will leak the dupped memory of "copy" (and result in memory corruption later) because the copy ptr is advanced and thus the the `_cleanup_free_` above destroy a pointer that points to the end of the actually allocated block

gportay replied:
That was a misunderstanding of https://github.com/systemd/systemd/pull/17576#discussion_r540191192

poettering commented on src/veritysetup/veritysetup.c:
```diff
@@ -43,6 +46,59 @@ static int help(void) {
+                        if (r < 0)
+                                return log_oom();
+                /* backward compatibility with obsolete ROOTHASHSIG positional argument */
+                } else if ((val = startswith(word, "base64:")) || stat(word, &st) == 0) {
```
hmm, use access(F_OK) to check if a file exists.

I'd really move this check into a helper function `looks_like_roothashsig()` or so.

And I figure this should use path_is_absolute() and is_path() to validate the argument (whch effectively means we'll loook at a leading "/"). We required the path to be absolute, so everything should be good.

poettering replied:
that said, I think this logic should be moved to the beginning of the function, before the extract_first_word_loop(). After all an option string of "/foo/bar/baz,waldo", should be detected as a signature file of `/foo/bar/baz,waldo`, and not one of `/foo/bar/baz` with an extra option `waldo`, for full compat with the old logic.

poettering commented on man/systemd.special.xml:
```diff
@@ -186,6 +186,13 @@
+            <para>A target that pulls in setup services for all
+            verity block devices.</para>
+          </listitem>
+        </varlistentry>
```
hmm, i wonder if we shouldn't just reuse cryptsetup.target for that. Do you have a strong reason to add a new target for this?

i.e. sure, it will be slightly weirdly named, but I'd keep things simple and say a single milestone for both verity and encrypted devices should suffice.

bluca replied:
Well fstab and crypttab have their own target (lots of other generators we supply do as well), so to me it would make sense to have one, so that it's clear they are related to different things.

gportay replied:
The drawback I see in creating a new target for `veritysetup` is the compatibility with what exists already.

The generated root units still reference to the previous target [cryptsetup-pre.target](https://github.com/systemd/systemd/blob/8374bf4fa2663b19f329801abfd8b29e66897380/src/veritysetup/veritysetup-generator.c#L100). Out of scope, maybe it should refer to `cryptsetup.target` directly?.

If we change the target it breaks the boot for people that uses initramfs because none of them include this newly introduced target `veritysetup.target`, and then, it creates a strong dependency between systemd and dracut for people using verity for root.

Anyway, I have some configuration files for dracut ready for that.

poettering replied:
i think distinct targets make sense if you want to handle what they group separately. I can see that for mounts and for crypt devices, since they are quite different, i.e. one is a thin kernel api, and the other requires a whole subsystem to work, device mgmt and everything. but the requirements for verity and cryptsetup are pretty much identical, i.e. is there any usecase where you#d want to manage the milestones for cryptsetup and veritysetups separately? I am not sure i see any...

I am not concerned about the dracut thing: we add new stuff all the time, dracut has to catch up, that's OK.

I am mostly interested about philosophically/conceptually if people ever want luks up but verity down or the opposite, or if they want verity up early and luks up late or so? i don't see that, do you?

poettering replied:
ah, well, let's leave this with distinct targets I guess. Maybe one day a usecase pops up, and it's better to have ti separate then already.

But please add counterparts for all of `cryptsetup-pre.target` + `cryptsetup.target` + `remote-cryptsetup.target`

bluca replied:
Speaking for my use case, I have a custom initramfs script that sets up verity and at some point in the near future I'd like to move to dracut + veritysetup generator. My disk grouped by cryptsetup.target is for external storage so to speak, not for the rootfs, and unlock is remote, so I could and in fact do have situations where my veritysetup targets need to be up (it's rootfs) and cryptsetup is not up yet or can't be up at all (first boot).

poettering reviewed (changes requested):
Looks great, some comments

poettering commented on src/veritysetup/veritysetup.c:
```diff
@@ -43,6 +46,81 @@ static int help(void) {
+#endif
+                else if ((val = startswith(word, "root-hash-signature="))) {
+
+                        r = free_and_strdup(&arg_root_hash_signature, val);
```
this should refuse values that are not an absolute path and not starting with "base64:"

gportay replied:
This should do the trick.

``` c
                        r = looks_like_roothashsig(val);
                        if (r < 0)
                                return r;
                        if (r == 0)
                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "root-hash-signature expects either full path to signature file or "
                                                                                "base64 string encoding signature prefixed by base64:.");
```

poettering commented on src/veritysetup/veritysetup.c:
```diff
@@ -43,6 +46,81 @@ static int help(void) {
+                return 0;
+        }
+
+        return 1;
```
return values appear reversed. The function names it primarily answers a boolean question (well it also cna return negative errors, but let's ignore that part), but then "yes" is returned as "0" and "no" as "1". That's unexpected... i would reverse the return codes

gportay replied:
Oops, indeed.

poettering commented:
Looks great. two nits above

gportay commented:
For information, I have fixed an issue with /usr (verity) mounted in the initrd. I will implement the remote-fs and push force afterward once it works well.

poettering commented:
veritysetup-pre.target is not documented in systemd.special(7) yet

poettering commented:
otherwise looks good.

(one of those days we should add /etc/integritytab, too ;-))

gportay commented:
> veritysetup-pre.target is not documented in systemd.special(7) yet

neither the `_netdev` option.

That is now fixed by  08b04ec.

bluca reviewed (approved):
LGTM - as far as I can see all comments have been fixed, so setting the green label. Waiting until at least Monday, to see if there are more follow-ups.

gportay commented:
@bluca: I wanted to test `root-hash-signature=` in a real situation, but I do not know how to proceed. is it hard to setup? What should I sign, the `roothash` in text/binary?

bluca commented:
> @bluca: I would like to test `root-hash-signature=` in a real situation, but I do not know how to proceed. is it hard to setup? What should I sign, the `roothash` in text/binary?

The hard part at the moment is that you need to have a key (or a cert that chains to a key) in the kernel trusted keyring, which is built-in. So most likely you have to build your own kernel. Work to allow using the secondary keyring too was merged in 5.11

Then, you sign the root hash value, in hex as it is printed by veritysetup. And then sign it with openssl smime.

eg:

```
veritysetup format img.raw img.verity | grep "Root hash" | awk '{print $3}' | tr -d '\n' > img.roothash
openssl smime -sign -in img.roothash -nocerts -inkey cert.key -signer cert.crt -noattr -binary -outform der -out img.roothash.p7s
```

gportay commented:
Thanks, I will give it a try.

gportay commented:
> veritysetup format img.raw img.verity | grep "Root hash" | awk '{print $3}' | tr -d '\n' > img.roothash
> openssl smime -sign -in img.roothash -nocerts -inkey cert.key -signer cert.crt -noattr -binary -outform der -out img.roothash.p7s

It works!

I have recompiled the kernel with the option `CONFIG_SYSTEM_TRUSTED_KEYS="cert.crt"`

```
[root@archlinux ~]# ls /dev/mapper/img 
/dev/mapper/img
[root@archlinux ~]# systemctl status systemd-veritysetup@img.service
● systemd-veritysetup@img.service - Integrity Protection Setup for img
     Loaded: loaded (8;;file://archlinux/etc/veritytab^G/etc/veritytab8;;^G; generated)
     Active: active (exited) since Fri 2021-01-15 17:24:08 EST; 53s ago
       Docs: 8;;man:veritytab(5)^Gman:veritytab(5)8;;^G
             8;;man:systemd-veritysetup-generator(8)^Gman:systemd-veritysetup-generator(8)8;;^G
             8;;man:systemd-veritysetup@.service(8)^Gman:systemd-veritysetup@.service(8)8;;^G
    Process: 462 ExecStart=/usr/lib/systemd/systemd-veritysetup attach img /root/img.raw /root/img.verity 112bf5bc9de3bfe00778>
   Main PID: 462 (code=exited, status=0/SUCCESS)

Jan 15 17:24:08 archlinux systemd[1]: Starting Integrity Protection Setup for img...
Jan 15 17:24:08 archlinux systemd-veritysetup[462]: Usage of ROOTHASHSIG positional argument is deprecated. Please use the opt>
Jan 15 17:24:08 archlinux systemd[1]: Finished Integrity Protection Setup for img.
```

bluca commented:
Nice! veritysetup should show some flag if the signature is loaded, and there's a kconfig to make it mandatory
